File: platforms/flutter/colyseus/lib/src/room_clock.dart
import 'dart:ffi';

import 'bindings/colyseus_core.dart';
import 'colyseus.dart';

/// Clock sync and RTT estimation for a room.
///
/// Fed by the TIMED prefix the server prepends to state messages once the room
/// declares inputs via `defineInput()`. A room that never declares inputs
/// never receives samples: [serverNow] then equals [now] and the rest read
/// zero.
///
/// Reach it through `room.clock`.
class RoomClock {
  final Pointer<colyseus_room_clock> _handle;

  RoomClock(this._handle);

  /// Local monotonic milliseconds — the base [serverNow] offsets.
  ///
  /// This is the timebase `predict.tick()` expects, so drive the frame loop
  /// from it rather than from `DateTime.now()`.
  double get now => core.colyseus_room_clock_now(_handle);

  /// Estimated server clock: milliseconds since the room started.
  double get serverNow => core.colyseus_room_clock_server_now(_handle);

  /// Server clock on a slew-limited timeline, for drawing.
  ///
  /// Free-runs at 1 ms/ms and servos toward [serverNow] instead of jumping, so
  /// interpolated entities don't stutter when an estimate lands. Never use it
  /// for server-stamped deadlines — use [serverNow].
  double get renderNow => core.colyseus_room_clock_render_now(_handle);

  /// Most recent round-trip sample in milliseconds; 0 before the first one.
  double get rtt => core.colyseus_room_clock_rtt(_handle);

  /// EMA-smoothed round-trip time; prefer this for forward prediction.
  double get smoothedRtt => core.colyseus_room_clock_smoothed_rtt(_handle);

  /// Interarrival jitter in milliseconds, against the advertised patch rate.
  double get jitter => core.colyseus_room_clock_jitter(_handle);

  /// Server encode time of the most recent snapshot.
  double get lastServerTime =>
      core.colyseus_room_clock_last_server_time(_handle);

  /// Advertised patch interval in milliseconds.
  double get patchInterval => core.colyseus_room_clock_patch_interval(_handle);

  /// The estimated offset between the local and server clocks.
  double get offset => core.colyseus_room_clock_offset(_handle);

  /// Time constant for [renderNow]'s slewing, in milliseconds.
  ///
  /// Lower reacts faster and jitters more; values <= 0 disable slewing so
  /// [renderNow] tracks [serverNow] exactly.
  set renderTau(double milliseconds) =>
      core.colyseus_room_clock_set_render_tau(_handle, milliseconds);

  /// The native handle.
  Pointer<colyseus_room_clock> get handle => _handle;
}

File: platforms/swift/Sources/Colyseus/RoomClock.swift
import CColyseus
import Foundation

public extension Colyseus {
    /// The room's sense of time.
    ///
    /// Local time is a monotonic millisecond count; server time is an estimate
    /// built from the timestamps the server stamps onto patches, and render
    /// time is that estimate slewed so it never jumps backwards.
    ///
    /// A room whose server never calls `defineInput()` gets no timestamps, and
    /// then ``serverNow`` equals ``now`` and the round-trip figures are zero.
    /// ``Colyseus/Room/ping()`` is the fallback there.
    struct RoomClock: @unchecked Sendable {
        // An opaque C struct, so Swift sees a bare pointer.
        let raw: OpaquePointer?

        init(_ raw: OpaquePointer?) {
            self.raw = raw
        }

        /// Local monotonic milliseconds. This is the timebase
        /// ``Colyseus/Predict/tick(_:)`` expects.
        public var now: Double { raw.map(colyseus_room_clock_now) ?? 0 }

        /// Estimated server time, in the server's own milliseconds.
        public var serverNow: Double { raw.map(colyseus_room_clock_server_now) ?? 0 }

        /// Server time, slew-limited — the timeline to interpolate remote
        /// entities along. Never runs backwards, even when an estimate
        /// correction would.
        public var renderNow: Double { raw.map(colyseus_room_clock_render_now) ?? 0 }

        /// The most recent round trip, in milliseconds.
        public var rtt: Double { raw.map(colyseus_room_clock_rtt) ?? 0 }

        /// Round trip with the spikes smoothed out. Prefer this on a HUD.
        public var smoothedRtt: Double { raw.map(colyseus_room_clock_smoothed_rtt) ?? 0 }

        /// Interarrival jitter, RFC 3550's estimator.
        public var jitter: Double { raw.map(colyseus_room_clock_jitter) ?? 0 }

        /// The server timestamp on the last patch. Watch it change to detect a
        /// patch arriving.
        public var lastServerTime: Double { raw.map(colyseus_room_clock_last_server_time) ?? 0 }

        /// Measured interval between patches, in milliseconds.
        public var patchInterval: Double {
            get { raw.map(colyseus_room_clock_patch_interval) ?? 0 }
            nonmutating set { raw.map { colyseus_room_clock_set_patch_interval($0, newValue) } }
        }

        /// How fast ``renderNow`` is allowed to catch up with a corrected
        /// estimate. Larger is smoother and further behind.
        public var renderTau: Double {
            get { 0 }
            nonmutating set { raw.map { colyseus_room_clock_set_render_tau($0, newValue) } }
        }

        /// Local-to-server offset the estimate currently believes.
        public var offset: Double { raw.map(colyseus_room_clock_offset) ?? 0 }
    }
}
